File: frontend/src/models/devicetree.ts
export type PropertyKind =
  | "boolean"
  | "string"
  | "string_list"
  | "cells"
  | "bytes"
  | "unknown";

export type PropertyValue = boolean | string | string[] | number[] | null;

export interface MetadataResponse {
  filename: string | null;
  file_size: number | null;
  node_count: number;
  property_count: number;
  warnings: string[];
  errors: string[];
}

export interface DeviceTreeProperty {
  name: string;
  raw_hex: string;
  kind: PropertyKind;
  value: PropertyValue;
}

export interface DeviceTreeNode {
  id: string;
  name: string;
  full_name: string;
  path: string;
  unit_address: string | null;
  parent_path: string | null;
  properties: DeviceTreeProperty[];
  children: DeviceTreeNode[];
}

export interface DeviceTreeResponse {
  node_count: number;
  root: DeviceTreeNode;
}

export interface ParseErrorDetail {
  source: string | null;
  warnings: string[];
  errors: string[];
}

export interface ErrorResponse {
  detail: ParseErrorDetail;
}

File: backend/app/api/schemas/devicetree.py
from __future__ import annotations

from pydantic import BaseModel

from app.model.devicetree import PropertyKind


PropertyValueResponse = bool | str | list[str] | list[int] | None


class MetadataResponse(BaseModel):
    filename: str | None
    file_size: int | None
    node_count: int
    property_count: int
    warnings: list[str]
    errors: list[str]


class DeviceTreePropertyResponse(BaseModel):
    name: str
    raw_hex: str
    kind: PropertyKind
    value: PropertyValueResponse


class DeviceTreeNodeResponse(BaseModel):
    id: str
    name: str
    full_name: str
    path: str
    unit_address: str | None
    parent_path: str | None
    properties: list[DeviceTreePropertyResponse]
    children: list["DeviceTreeNodeResponse"]


class DeviceTreeResponse(BaseModel):
    node_count: int
    root: DeviceTreeNodeResponse


class ParseErrorDetail(BaseModel):
    source: str | None
    warnings: list[str]
    errors: list[str]


class ErrorResponse(BaseModel):
    detail: ParseErrorDetail

